Userland/Libraries/LibSQL/Heap.cpp

```diff
@@ -66,118 +69,137 @@ ErrorOr<void> Heap::open()
-    }
-    if (buffer.size() > BLOCK_SIZE) {
-        warnln("Heap({})::write_block({}): Oversized block ({} > {})"sv, name(), block, buffer.size(), BLOCK_SIZE);
-        return Error::from_string_literal("Heap()::write_block(): Oversized block");
-    }
+    dbgln_if(SQL_DEBUG, "Write raw block {}", index);
 
-    dbgln_if(SQL_DEBUG, "Write heap block {} size {}", block, buffer.size());
-    TRY(seek_block(block));
+    VERIFY(m_file);
+    VERIFY(data.size() == Block::SIZE);
 
-    if (auto current_size = buffer.size(); current_size < BLOCK_SIZE) {
-        TRY(buffer.try_resize(BLOCK_SIZE));
-        memset(buffer.offset_pointer(current_size), 0, BLOCK_SIZE - current_size);
-    }
+    TRY(m_file->seek(index * Block::SIZE, SeekMode::SetPosition));
+    TRY(m_file->write_until_depleted(data));
 
-    dbgln_if(SQL_DEBUG, "{:hex-dump}", buffer.bytes().trim(8));
-    TRY(m_file->write_until_depleted(buffer));
+    if (index > m_highest_block_written)
+        m_highest_block_written = index;
 
-    if (block == m_end_of_file)
-        m_end_of_file++;
     return {};
 }
 
-ErrorOr<void> Heap::seek_block(u32 block)
+ErrorOr<void> Heap::write_raw_block_to_wal(Block::Index index, ReadonlyBytes data)
 {
-    if (!m_file) {
-        warnln("Heap({})::seek_block({}): Heap file not opened"sv, name(), block);
-        return Error::from_string_literal("Heap()::seek_block(): Heap file not opened");
-    }
-    if (block > m_end_of_file) {
-        warnln("Heap({})::seek_block({}): Cannot seek beyond end of file at block {}"sv, name(), block, m_end_of_file);
-        return Error::from_string_literal("Heap()::seek_block(): Cannot seek beyond end of file");
-    }
+    dbgln_if(SQL_DEBUG, "{}(): adding raw block {} to WAL", __FUNCTION__, index);
+    VERIFY(index < m_next_block);
+    VERIFY(data.size() == Block::SIZE);
 
-    if (block == m_end_of_file)
-        TRY(m_file->seek(0, SeekMode::FromEndPosition));
-    else
-        TRY(m_file->seek(block * BLOCK_SIZE, SeekMode::SetPosition));
+    auto buffer = TRY(ByteBuffer::copy(data));
+    TRY(m_write_ahead_log.try_set(index, move(buffer)));
 
     return {};
 }
 
-u32 Heap::new_record_pointer()
+ErrorOr<void> Heap::write_block(Block const& block)
 {
-    VERIFY(m_file);
-    if (m_free_list) {
-        auto block_or_error = read_block(m_free_list);
-        if (block_or_error.is_error()) {
-            warnln("FREE LIST CORRUPTION");
-            VERIFY_NOT_REACHED();
-        }
-        auto new_pointer = m_free_list;
-        memcpy(&m_free_list, block_or_error.value().offset_pointer(0), sizeof(u32));
-        update_zero_block();
-        return new_pointer;
-    }
-    return m_next_block++;
+    VERIFY(block.index() < m_next_block);
+    VERIFY(block.next_block() < m_next_block);
+
+    auto size_in_bytes = block.size_in_bytes();
+    auto next_block = block.next_block();
+
+    auto heap_data = TRY(ByteBuffer::create_zeroed(Block::SIZE));
```

Comment:
We can live with it for now, but all the re-allocations of the byte buffers involved jump out to me. Would be nice if we can avoid these reallocs in the future.

Reply:
Yeah I could probably be a bit smarter about these. I've found a way to not allocate anything when writing to the WAL, so I included that in this PR.